### PR TITLE
feat: add ability to save an assistant

### PR DESCRIPTION
**Pull request by thedadams**
If a user has added tools to a thread, then they will be able to save the assistant such that these tools become part of the assistant. The only exception to this is the knowledge tool: the user can't explicitly add the knowledge tool, so it will not be saved to the assistant. This will change when a user has the ability to add knowledge to an assistant.

### CODE REVIEW

File: contexts/chat.tsx [hunk 1]
```diff
@@ -31,8 +31,8 @@ interface ChatContextState {
   setSubTool: React.Dispatch<React.SetStateAction<string>>;
   setScript: React.Dispatch<React.SetStateAction<string>>;
   setScriptContent: React.Dispatch<React.SetStateAction<Block[] | null>>;
-  tool: Tool;
-  setTool: React.Dispatch<React.SetStateAction<Tool>>;
+  rootTool: Tool;
```
Comment (thedadams) [inline contexts/chat.tsx:34]: I keep getting confused that this is only the "root tool" and not the whole tool, so I renamed it. Happy to switch it back if folks disagree with this change.

Comment (g-linville) [reply contexts/chat.tsx:34]: I like this

File: components/scripts/script-save.tsx [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+  const knowledgeTool = gatewayTool();
+
+  async function saveScript() {
+    if (!rootTool.tools || rootTool.tools.length === 0) {
+      rootTool.tools = [];
+    }
+
+    // The knowledge tool is dynamic and not controlled by the user. Don't add it to the saved tool.
+    const addedTools = tools.filter((t) => t !== knowledgeTool);
```
Comment (StrongMonkey) [inline components/scripts/script-save.tsx:27]: 👍

**Review by StrongMonkey — APPROVED** [review]

File: components/scripts/script-save.tsx [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+  return (
+    <Dropdown placement="bottom" closeOnSelect={false}>
+      <DropdownTrigger>
+        <Button variant="light" isIconOnly>
+          <PiFloppyDiskThin className="size-5" />
+        </Button>
+      </DropdownTrigger>
+      <DropdownMenu>
+        {scriptId ? (
+          <DropdownItem key="save" onPress={() => saveScript()}>
```
Comment (g-linville) [inline components/scripts/script-save.tsx:56]: nit: I think this works the same:

```suggestion
          <DropdownItem key="save" onPress={saveScript}>
```

**Review by g-linville — APPROVED** [review]

File: components/scripts/script-save.tsx (hunk 3 above)
Comment (thedadams) [reply components/scripts/script-save.tsx]: Done.

**Review by njhale — APPROVED** [review]

**Comment by thedadams** [comment]
In my drive, I realized there is a small issue with this. I will fix it tonight.

**Review by ryanhopperlowe — APPROVED** [review]

**Pull request merged**